src/renderer/cursor_renderer/mod.rs

```diff
@@ -341,6 +347,54 @@ impl CursorRenderer {
 
         canvas.restore();
 
+        // draw ime if possible under current cursor
+        if let Some(preedit) = preedit {
+            let factor = grid_renderer.font_dimensions.width as f32;
+            let y_adjustment = y_adjustment as f32 + factor / 5.0;
+
+            let off = if let Some(off) = preedit.cursor_offset() {
+                off as f32 * factor
+            } else {
+                0f32
+            };
+
+            let mut start = self.destination.x;
+            let mut end = self.destination.x + off;
+            if end + factor > w {
+                start = start - (end + factor - w);
+                end = w - factor;
+            }
+
+            use skia_safe::Rect;
+            let r = Rect::new(
+                start,
+                self.destination.y,
+                end,
+                self.destination.y + y_adjustment + factor / 4.0,
+            );
+
+            paint.set_color(background_color);
+            canvas.draw_rect(r, &paint);
+
+            let blobs =
+                &grid_renderer
+                    .shaper
+                    .shape_cached(preedit.preedit_text().clone(), bold, italic);
+            paint.set_color(foreground_color);
+            for blob in blobs.iter() {
+                canvas.draw_text_blob(blob, (start, self.destination.y + y_adjustment), &paint);
+            }
+
+            paint.set_color(skia_safe::colors::GREY.to_color());
```

Comment:
This should use the correct highlight group color, otherwise it might be invisible. I think this is the selection right?

Also, shouldn’t it be drawn before the text? So that the text remain invisible, with the text color also set apporopriately through the highlight groups.

I also think that the floating window transparency need to be taken into account.